xla/service/gpu/runtime/nccl_collective_permute_thunk.h

```diff
@@ -46,26 +50,31 @@ class NcclCollectivePermuteStartThunk : public NcclCollectiveThunk {
   NcclCollectivePermuteStartThunk(ThunkInfo thunk_info, NcclApi* nccl_api,
                                   const HloCollectivePermuteInstruction* instr,
                                   int64_t replica_count,
-                                  int64_t partition_count,
-                                  const Buffer& buffer);
+                                  int64_t partition_count, const Buffer& buffer,
+                                  bool p2p_memcpy_enabled);
+  absl::Status Initialize(const InitializeParams& params) override;
 
   static const char* GetHloOpName() { return "collective-permute-start"; }
 
  protected:
   const NcclCollectiveConfig& config() const override { return config_.config; }
   absl::Status RunNcclCollective(const ExecuteParams& params,
                                  se::Stream& stream,
-                                 NcclApi::NcclCommHandle comm) override;
+                                 NcclApi::NcclCommHandle comm,
+                                 bool is_local) override;
 
  private:
   const NcclP2PConfig config_;
   const Buffer buffer_;
+  absl::node_hash_map<int64_t, AsyncValueRef<void*>> recv_ptr_map_;
```

Comment:
This map has to be guarded with a mutex because thunks run concurrently, also you pass it around by reference to few function, I'd do something like AsyncEvents to wrap the map and add synchronization: https://github.com/openxla/xla/blob/main/xla/service/gpu/runtime/nccl_collective_thunk.h#L133-L148

Reply:
I agree we need to lock when doing initialization since we only do rendezvous after initialization in gpu_executable. Once we properly initialize the map, we shouldn't need more locking there? the current_id is unique, so each thread will only modify their own entry in the map and the synchronization is ensured by the value which is an AsyncValueRef, the sending thread will not proceed until the receiving thread signals that the value is in available state.

Reply:
Yes, looks like it should be safe, it's just not clear from the function signature because you pass mutable map around

Reply:
Also I'd add an alias for this map name, it's passed around in a few places, and it's very long type name